hardware/smart_utils.py

```diff
@@ -0,0 +1,263 @@
+    return hwlst
+
+
+def read_smart_ata(hwlst, device, optional_flag="", mode=""):
+    foundid = False
+    device_name = os.path.basename(device)
+    optional_string = ""
+    if optional_flag:
+        optional_string = " with %s" % optional_flag
+
+    values = {}
+    if mode:
+        device_name = "%s{%s}" % (device_name, optional_flag.split()[1])
+
+    sdparm_cmd = subprocess.Popen("smartctl -a %s %s" % (device,
+                                                         optional_flag),
+                                  shell=True, stdout=subprocess.PIPE)
+    for line in sdparm_cmd.stdout:
+        line = line.strip()
+        if isinstance(line, six.binary_type):
+            line = line.decode()
+
+        if read_smart_field(hwlst, line, device_name, "Device Model:",
+                            "device_model"):
+            sys.stderr.write("read_smart_ata: Found S.M.A.R.T information "
+                             "on %s%s\n" % (device, optional_string))
+            continue
+
+        if read_smart_field(hwlst, line, device_name, "Serial Number:",
+                            "serial_number"):
+            continue
+
+        if read_smart_field(hwlst, line, device_name, "Firmware Version:",
+                            "firmware_version"):
+            continue
+
+        if line.startswith("ID#"):
+            foundid = True
+            continue
+        if foundid is False:
+            continue
+        elif not line:
+            break
+        try:
+            fields = line.split()
+            if len(fields) < 10:
+                raise ValueError('Expected at least 10 fields in %s, found %d.' %
+                                 (line, len(fields)))
+            values["id"] = fields[0]
+            values["name"] = fields[1]
+            values["flag"] = fields[2]
+            values["value"] = fields[3]
+            values["worst"] = fields[4]
+            values["thresh"] = fields[5]
+            values["type"] = fields[6]
+            values["updated"] = fields[7]
+            values["when_failed"] = fields[8]
+            if values["when_failed"] == "-":
+                values["when_failed"] = "NEVER"
+            raw_values = fields[9:]
+            raw_value = ""
+            for raw in raw_values:
+                raw_value = "%s %s" % (raw_value, raw)
+            values["raw"] = raw_value
+            for title in ["value", "worst", "thresh", "when_failed", "raw"]:
+                hwlst.append(("disk", device_name,
+                              "SMART/%s(%s)/%s" % (values["name"],
+                                                   values["id"],
+                                                   title),
+                              values[title]))
+            continue
+
+        except Exception:
+            sys.stderr.write("read_smart: failed to read line : %s\n" % line)
+            continue
+
+
+def read_smart(hwlst, device, optional_flag=""):
+    if not which("smartctl"):
```

Comment:
The reason this is failing locally is I don't have a smartctl so this 'which' call is returning None.  and therefore the rest of the test isn;t run resulting in the fail.

Please also mock this out in test_smart_utils.py

Reply:
done